### ISSUE TITLE
Additional Windows

### ISSUE THREAD

**Issue by pigmalien** [Scirra/Construct-feature-requests#64]
### Reviewed guidelines

- [X] I have read and understand the suggestion guidelines

### Checked for duplicate suggestions

- [X] I checked for existing similar suggestions

### Summary

To better deal with a crowded editor, and allow a sandboxed development platform for third party access to a visual layout editor, IE user created interfaces for the layout editor, I propose a slave type window that can communicate with the main editor, as well as offer the viability of multiple windows, and monitors.

### Possible workarounds or alternatives

While its possible to drag a few of the dialogs from the editor to other monitors, it is very limited.

### Proposed solution

The main editor would have an option to create a new window or tab that can be dragged to where ever the user prefers. Those windows might offer the ability to drag existing tools from the editor to use in the new window, or just offer new official workspace, as well as giving the third party developers a way to create tools that aren't plausible due to security concerns. The main layout would most likely still be completely insolated. It would essentially work like a html element rendered in its own window to allow existing JavaScript tools like Angular, Svelt, etc. set up by third parties.

### Why is this idea important?

We need more room. We need third party access to layout tools, besides just a third party editor.

### Additional remarks

It wouldn't have to a fast synch wise, but hopefully new methods make it easier now than before.

**Comment by OverboyDev** [comment]
Multiple screen window management is extensively supported by Chrome

Demo: https://michaelwasserman.github.io/window-placement-demo/
Chrome feature: https://chromestatus.com/feature/5252960583942144

It would be great to implement this properly in Construct (multiple popup, store/retrieve Custom Editor Layouts), even if only Chrome is supported yet (need to check browser compat) as it wouldn't matter for players anyway.

(if it's not supported yet by Firefox/Safari, it could just keep the existing way it works for them but it shouldn't prevent all C3 users to be able to create their own editor layout as most game engine allow to do)

**Comment by Jase0000** [comment]
Politely, maybe I'm missing something, but aren't most parts of C3 able to be dragged onto different monitors?

Find, References, Bookmarks, Event Sheets, Layouts, Properties, Objects, Layers, Tilemap, Timeline, all allow popup windows to drag to other monitors.

Funnily enough I was thinking to suggest (if possible) a way to remember this, as it's a pain to have to have to pop out each pane upon opening C3.

I have no comment on the third party stuff, not understood the benefits, but assuming there are benefits, I'm eager to see workflows improved!

**Comment by LuanHimmlisch** [comment]
(Sorry for the large comment, I got myself carried away lol)

This suggestion tackles two points:
- Extending the Editor SDK.
- Being able to pop-out Construct windows.

As for extending the SDK, I don't think that actual popup windows are necessary, they could work like the current Construct windows. However, as mentioned, it would be a security concern, as they would be able to access the user's editor context. So popups are a solution for isolation, besides accomplishing the second point of giving the option of better multi-monitor support.

However, popups complicate the development quite a bit. And I'm totally against developing them on an experimental spec like the one Overboy mentions.  

But window management already exists on the web, and Construct already does it with the preview window using [postMessages](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage). It's just a matter of doing the same with every window (better said than done). This would require making a whole API of tailor-made endpoints for any action you would ever need, and I'm sure it would be a very big thing to tackle.

Even if we go without the popups and use Construct windows, I can't think of any cheaper solution besides one:

I imagine an SDK configuration where we list each window:

```js
this._info.SetWindows([
    {
        id: "",
        html: "./window.html",
        style: "./style.html",
        functions: {
            myFunction: `
            alert("I'm isolated using a parser")
            `;
        }
    },
])
```

The fetched HTML should be [sanitized](https://www.npmjs.com/package/sanitize-html) to prevent any JavaScript.

The HTML would allow a special attribute to reference a function registered, something like this:

```
<button data-onclick="myFunction">Add sprite</button>
```

Then it would register the events and on execution it would check if the function is registered and eval the string using a [parser](https://www.npmjs.com/package/@tootallnate/quickjs-emscripten).

So basically, instead of using iFrames, Windows or Web Workers that restrict everything, we manually parse both the HTML and the JS.

This option is a bit cheaper since, instead of making a whitelist, building from scratch an API; we are just blacklisting the already existing functions the editor uses.

**Comment by OverboyDev** [comment]
@LuanHimmlisch The Chrome Window Management feature was release in Chrome 93 and is considered as stable since Chrome 100. We're currently at Chrome 120, it's been stable for 2 years.

(But it looks like you know a lot more about webdev than I do so maybe I miss something)

**Comment by LuanHimmlisch** [comment]
@OverboyDev yeah my comment was about the other browsers. I understand that, as a customer, you may not care about them, but still, the web must be built upon accepted standards (You can check the support of that spec [here](https://caniuse.com/?search=getScreenDetails)). 

Nonetheless, my comment was hyperbolic, since the spec is not about a new way of creating windows (something I assumed since I overlooked the spec hehe), it's about building upon the already functional spec, adding access to some additional info.

So if the suggestions go full-through, with the popups and everything, I imagine Chrome users may benefit from some extra monitor functionality, while other browsers will have the same ol' windows 😄

**Comment by AshleyScirra** [comment]
It's not clear whether this suggestion is about being able to move existing windows/panes to other windows (which should already be supported by most panes and editors, and some dialogs), or being able to make custom editor windows in the SDK (a hugely complex task as it would need a very big API surface to do anything useful, especially if it is to use the same UI controls/style as the rest of the editor). That also means it's not especially clear which people are voting on.

To clear up the ambiguity, I'm closing this feature request as declined - please file a new suggestion focused on the particular area you meant.
